crates/wasmparser/src/validator/operators.rs

```diff
@@ -1625,6 +1649,17 @@ where
         };
         Ok(())
     }
+    fn visit_global_atomic_get(
+        &mut self,
+        _ordering: crate::Ordering,
+        global_index: u32,
+    ) -> Self::Output {
+        self.visit_global_get(global_index)?;
+        // No validation of `ordering` is needed because `global.atomic.get` can
+        // be used on both shared and unshared globals.
+        self.check_atomic_global_ty(global_index, Some(RefType::ANYREF.into()))?;
```

Comment:
Ok so apparently I was confused about this and sorry if that caused issues.

I only just now realize that the intention is to be able to use `global.atomic.get` with non-shared globals in addition to shared globals. If that's the goal then that means that the type-checking code is duplicated both at shared-global-definition-time but additionally here.

Spec-level questions for you:

* Could the spec be stricter and say that `global.atomic.*` instructions only work on `shared` globals? IIRC the rationale for `memory.atomic.*` working on non-atomic memories isn't quite as applicable to globals since globals, ironically, aren't as "global" as memory since there's only one linear memory but often many globals.
* Failing that, could the spec allow any type with non-shared globals and `global.atomic.*`?
* Failing that, mind refactoring the validation here to have a shared helper that verifies that the type used here is valid for atomic operations? That would then be shared with the validation of a global declaration.

Also, an orthogonal spec-level question: is it valid to `global.get` a shared global? Mostly want to confirm if preexisting instructions need to be updated in their validation.

Reply:
> is it valid to global.get a shared global?

Yes, it's the way to do an unordered access.

> Could the spec be...?

I think we confused each other here. I was a little too in the weeds figuring out how to do all of this in wasm-tools and didn't see the big picture: actually, defining a global &mdash; shared or otherwise &mdash; can be done with _any_ possible type. E.g., someone might want to store a `v128` in a shared global or a `funcref` or a float... We shouldn't really be putting any extra validation on the global definition.

The values in those shared globals can be accessed as usual by `global.get|set` in an unordered way. It's only if we want to enforce atomic access that we limit the types to `i32 | i64 | <: anyref`. So we really only need the extra validation there.

For context From [here](https://github.com/WebAssembly/shared-everything-threads/blob/main/proposals/shared-everything-threads/Overview.md#globals)

> The new instructions are valid for both shared and unshared globals.

Reply:
See c15c281.